merkle/src/tree.rs

```diff
@@ -0,0 +1,22 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0
+
+/// Trait for implementing general n-degree merkle tree
+pub trait MerkleTree<const DEGREE: usize> {
+    /// Type of the leaf data
+    type Item;
+    /// Type of the hash values
+    type Hash;
+
+    /// Depth of the tree
+    fn depth(&self) -> u32;
+    /// Number of leafs
+    fn count(&self) -> usize;
+    /// Root hash, lazy-evaluated
+    fn root(&mut self) -> Option<Self::Hash>;
+
+    /// Adds a new leaf
+    fn add(&mut self, item: Self::Item);
+    /// Adds a batch of leaves in the give order
+    fn add_batch(&mut self, items: Vec<Self::Item>);
```

Comment:
I think requiring a function to accept a vector isn't the best practice. Instead something like
```rust 
fn add_batch<I: IntoIterator<Self::Item>>(&mut self, items: I);
```
would be better. This would eliminate a bunch of conversions to vectors and let us pass arrays, slices and other things

Reply:
Fixed